src/v/cluster/plugin_frontend.h

```diff
@@ -0,0 +1,164 @@
+      = std::variant<transform_update_cmd, transform_remove_cmd>;
+
+public:
+    // The plugin frontend takes the corresponding services for it's core.
+    //
+    // The exception being the controller_stm which is only nonnull for the
+    // service on core 0.
+    plugin_frontend(
+      model::node_id,
+      partition_leaders_table*,
+      plugin_table*,
+      topic_table*,
+      controller_stm*,
+      rpc::connection_cache*,
+      ss::abort_source*);
+
+    using notification_id = plugin_table::notification_id;
+    using notification_callback = plugin_table::notification_callback;
+    struct mutation_result {
+        // The UUID of the modified transform, mostly useful for cleanup
+        // of wasm_binaries.
+        // See transform_metadata::uuid
+        uuid_t uuid;
+        errc ec;
+    };
+
+    // Create or update a transform by name.
+    ss::future<errc> upsert_transform(
+      model::transform_metadata, model::timeout_clock::time_point);
+
+    // Remove a transform by name.
+    ss::future<mutation_result>
+      remove_transform(model::transform_name, model::timeout_clock::time_point);
+
+    // Register for updates going forward.
+    //
+    // This is a core local operation, you can register for updates to the
+    // transforms table. If you want to lookup the transform you can then do
+    // that in the callback.
+    notification_id register_for_updates(notification_callback);
+
+    // Unregister a previously registered notification.
+    void unregister_for_updates(notification_id);
+
+    // Lookup a transform by ID. Useful when paired with registering for
+    // updates.
+    std::optional<model::transform_metadata>
+      lookup_transform(model::transform_id) const;
+
+    // Lookup transforms for input topics.
+    absl::flat_hash_map<model::transform_id, model::transform_metadata>
+      lookup_transforms_by_input_topic(model::topic_namespace_view) const;
+
+    // Get a snapshot of all the transforms that currently exist.
+    absl::flat_hash_map<model::transform_id, model::transform_metadata>
+    all_transforms() const;
+
+private:
+    // Perform a mutation request, check if this node is the cluster leader and
+    // ensuring that the right core is used for local mutations.
+    ss::future<mutation_result>
+      do_mutation(transform_cmd, model::timeout_clock::time_point);
+
+    // If this node is not the leader, dispatch this mutation to `node_id` as
+    // the cluster's leader.
+    ss::future<mutation_result> dispatch_mutation_to_remote(
+      model::node_id, transform_cmd, model::timeout_clock::duration);
+
+    // Performs (and validates) a mutation command to be inserted into the
+    // controller log.
+    //
+    // This must take place on the controller shard on the cluster leader with
+    // _mu held to ensure consistency.
+    ss::future<mutation_result>
+      do_local_mutation(transform_cmd, model::timeout_clock::time_point)
+        ABSL_EXCLUSIVE_LOCKS_REQUIRED(_mu);
```

Comment:
is this due to foreign thread? why do we need a lock

Reply:
i mean an old school lock. not a seastar lock.

Reply:
This is a seastar lock. The annotation is mostly documentation, but I'm playing around to see if clang can do thread safety analysis with coroutines: https://clang.llvm.org/docs/ThreadSafetyAnalysis.html

Reply:
i think it's a seastar lock, but this is adding the `exclusive_locks_required` attribute that clang understand.

but... does that really help us? it seems like clang would be assuming a threaded model that isn't relevant here...

Reply:
1) It's for documentation
2) clang's analysis here is entirely lexical analysis, it doesn't actually run the code or look at multiple threads. Multiple fibers are practically the same from a locking control flow mechanism. All the "thread safety mechanisms" in clang do is make sure a function is called within that scope.

I can remove these if we find them more confusing than helpful. We don't add the compiler flag to enable thread safety so these don't do anything. I'm playing around in godbolt to see if it's possible to get these to work in coroutines

Reply:
https://godbolt.org/z/ocdshhred

It seems to work fine in coroutines, but I can't figure out how to get RAII to work with it... Here's the absl verison: https://github.com/abseil/abseil-cpp/blob/197bac8574c511b1937b60ce6405462f9af678ce/absl/synchronization/mutex.h#L544

Reply:
>I can remove these if we find them more confusing than helpful.

if they are helpful then i think it's just a matter of learning about it. so if it providing value and it sounds like it is working with coroutines which is pretty cool then maybe we can just have a 5 minutes slot at core panda to disseminate the information?